File: nc_150/binary_search/time_based_key_val_store.cpp
// idea is to make use of map of vectors which contains pairs of  val, timestamp
// as this stores keys, so something related to map has to be used here
// As the timestamps are strictly increasing, so we can apply binary search on the 
// timestamps and we can return the value corresponding to the key given accordingly
// TC = O(logn)

class TimeMap {
public:
    unordered_map<string, vector<pair<string,int>>>m;
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        m[key].push_back({value, timestamp});
    }
    
    string get(string key, int timestamp) {
        if(m.find(key) == m.end())
            return "";

        return search(m[key], timestamp);
    }

    string search(vector<pair<string,int>>&vec, int &ts){
        int l = 0, r = vec.size();
        string res = "";
        while (l < r){
            int mid = l + (r-l)/2;
            if (vec[mid].second < ts){
                l = mid + 1;
                res = vec[mid].first;
            } else if (vec[mid].second > ts)
                r = mid ;
            else 
                return vec[mid].first;
        }
        return res;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */

File: nc_150/binary_search/go codes/time_based_key_val_store.go
// we will make use of the condition given in the contraints:
// that the timestamps of the set are strictly increasing
// we will simply apply binary search to find the nearest timestamp for the given key
// and accordingly return the val

type Pair struct {
	value     string
	timestamp int
}

type TimeMap struct {
	mp map[string][]Pair
}

func Constructor() TimeMap {
	return TimeMap{make(map[string][]Pair)}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	if _, ok := this.mp[key]; !ok {
		this.mp[key] = make([]Pair, 0)
	}
	this.mp[key] = append(this.mp[key], Pair{value, timestamp})
}

func (this *TimeMap) Get(key string, timestamp int) string {
	vals, ok := this.mp[key]
	if !ok {
		return ""
	}

	var val string
	l, r := 0, len(vals)-1

	for l <= r {
		mid := l + (r-l)/2
		if vals[mid].timestamp <= timestamp {
			val = vals[mid].value
			l = mid + 1
		} else {
			r = mid - 1
		}
	}
	return val
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */